### PR TITLE
ARQ-1902 Provide SPI for conditional execution of test methods

### PR DESCRIPTION
**Pull request by smiklosovic**
@aslakknutsen 

I am not sure if you want it with that ThreadLocal version, I have added it there according to this (1). Just cherry-pick what version is more suitable for you and evenutally squash it.

ThreadLocal version also does kind of caching ... so we do not load that decider via manager every time it is called and it is resolved by manager only once.

(1) https://github.com/arquillian/arquillian-core/blob/master/test/spi/src/main/java/org/jboss/arquillian/test/spi/TestRunnerAdaptor.java#L24

### CODE REVIEW

**Comment by smiklosovic** [comment]
dont merge it, i have some issues here

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java [hunk 1]
```diff
@@ -48,6 +51,8 @@
 {
    private Manager manager;
    
+   private static final ThreadLocal<TestExecutionDecider> testExecutionDecider = new ThreadLocal<TestExecutionDecider>();
```
Comment (aslakknutsen) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: Avoid any caching here. Resolve it from the ServiceLoader each time, let it cache if required. The service should use the contexts etc if it require other resources via injection.

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java [hunk 2]
```diff
@@ -126,4 +154,26 @@ public void shutdown()
...
+       {
+           return cachedTestExecutionDecider;
+       }
+
+       Validate.notNull(manager, "Manager must be specified.");
+       ServiceLoader serviceLoader = manager.resolve(ServiceLoader.class);
+
+       if (serviceLoader != null)
+       {
+           TestExecutionDecider decider = serviceLoader.onlyOne(TestExecutionDecider.class, DefaultTestExecutionDecider.class);
```
Comment (aslakknutsen) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: Why restrict to only one impl? 

EXECUTE if all Deciders agree?

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java [hunk 3]
```diff
@@ -108,6 +127,15 @@ public TestResult test(TestMethodExecutor testMethodExecutor) throws Exception
       Validate.notNull(testMethodExecutor, "TestMethodExecutor must be specified");
 
       final List<TestResult> result = new ArrayList<TestResult>();
+      
+      TestExecutionDecider testExecutionDecider = resolveTestExecutionDecider(manager);
+
+      if (testExecutionDecider != null && !testExecutionDecider.execute(testMethodExecutor.getMethod()))
+      {
+          result.add(TestResult.skipped(null));
```
Comment (aslakknutsen) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: In the JUnit integration we actually only check if testResult.hasException() {throw Exception}.. If the test is skipped the JUnit specific exception AssumptionViolatedException is caught from JUnit and rethrown.. We would need to update the JUnit integration as well to check for the State==skipped and recreate a AssumptionViolatedException

File: test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/TestExecutionDecider.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+    };
+
+    /**
+     * 
+     * @param testMethod test method to resolve a test execution on
+     * @return true if execution is performed, false otherwise
+     */
+    boolean execute(Method testMethod);
```
Comment (aslakknutsen) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/TestExecutionDecider.java]: Wonder if more then just a bool would be apprichiated here. .e.g. a Cause Message for false. 

Somewhere we might want to say something about; who stopped it from executing and why.

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java (hunk 2 above)
Comment (kpiwko) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: :+1: , I don't think we need to model OR condition - if so it should be modelled in one decider

Comment (aslakknutsen) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: Agree

**Comment by smiklosovic** [comment]
@aslakknutsen I have added that execution decision, I am not sure what you want to do with that message, eventually log it before we return from the execution?

File: test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/TestExecutionDecider.java (hunk 4 above)
Comment (smiklosovic) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/TestExecutionDecider.java]: implemented in next commit

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java (hunk 2 above)
Comment (smiklosovic) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: @aslakknutsen so lets order these deciders according to precedence and do some chain of it

**Comment by aslakknutsen** [comment]
@smiklosovic This becomes a tiny bit hacky, but.. I think you need a ExecutionDecision to Exception converter of some sort. Then add that Exception to the TestResult. Later here; https://github.com/arquillian/arquillian-core/blob/master/junit/core/src/main/java/org/jboss/arquillian/junit/Arquillian.java#L342 you'll need to check for that type and convert it to the Junit Specific 'Skip' exception

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/execution/DefaultTestExecutionDecider.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import java.lang.reflect.Method;
+
+import org.jboss.arquillian.test.spi.execution.ExecutionDecision;
+import org.jboss.arquillian.test.spi.execution.TestExecutionDecider;
+
+/**
+ * @author <a href="mailto:smikloso@redhat.com">Stefan Miklosovic</a>
+ *
+ */
+public class DefaultTestExecutionDecider implements TestExecutionDecider
```
Comment (aslakknutsen) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/execution/DefaultTestExecutionDecider.java]: With support for multiple TestExecutionDecider this one is no longer needed. If non are found it should be EXECUTE.

File: test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    private Decision decision;
+
+    private String reason;
+
+    public static ExecutionDecision execute()
+    {
+        return new ExecutionDecision(Decision.EXECUTE);
+    }
+
+    public static ExecutionDecision execute(String reason)
```
Comment (aslakknutsen) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java]: execute is the default behavior, so not sure we need a Reason for not doing aything.

**Comment by smiklosovic** [comment]
@aslakknutsen yeah I am on it ... nice example how things can easilly get complicated :)

File: test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        return new ExecutionDecision(Decision.EXECUTE, reason);
+    }
+
+    public static ExecutionDecision dontExecute()
```
Comment (aslakknutsen) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java]: Don't allow to _not_ give a reason when skipping

File: test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java (hunk 6 above)
Comment (smiklosovic) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java]: was thinking about it, but implementators of that deciders could indeed put there some reason why we are going to execute it anyway

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java (hunk 2 above)
Comment (kpiwko) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: Let's making a chain of responsibility, passing latest ExecutionDecision to next ExecutionDesider sorted by priority :-)

**Comment by aslakknutsen** [comment]
@smiklosovic Similar, somewhere here https://github.com/arquillian/arquillian-core/blob/master/testng/core/src/main/java/org/jboss/arquillian/testng/Arquillian.java#L218 you'll need either to create a SkipException or set setStatus(skip) on the TestNG TestResult object. Not 100% sure which it is. A quick look at the code seems to imply both will work but.. :)

**Comment by smiklosovic** [comment]
nice, so right now it is marked as skipped from junit :) only testng is missing

**Comment by smiklosovic** [comment]
@aslakknutsen considered done

File: test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java [hunk 8]
```diff
@@ -108,6 +128,15 @@ public TestResult test(TestMethodExecutor testMethodExecutor) throws Exception
       Validate.notNull(testMethodExecutor, "TestMethodExecutor must be specified");
 
       final List<TestResult> result = new ArrayList<TestResult>();
+
+      ExecutionDecision executionDecision = resolveExecutionDecision(manager, testMethodExecutor.getMethod());
+
+      if (executionDecision.getDecision() == Decision.DONT_EXECUTE)
+      {
+          result.add(TestResult.skipped(new Exception(executionDecision.getReason())));
```
Comment (aslakknutsen) [inline test/impl-base/src/main/java/org/jboss/arquillian/test/impl/EventTestRunnerAdaptor.java]: This should be new SkippedTestExecutionException no?

File: test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/TestExecutionDecider.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        @Override
+        public int precedence() {
+            return 1;
+        }
+
+    };
+
+    static final TestExecutionDecider DONT_EXECUTE = new TestExecutionDecider()
```
Comment (aslakknutsen) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/TestExecutionDecider.java]: Wondering if we should remove this as it entices developers to return a 'non sense' reason..

Comment (smiklosovic) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/TestExecutionDecider.java]: yes I can delete it, it is used only in tests and I can declare it there

File: test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java (hunk 6 above)
Comment (aslakknutsen) [inline test/spi/src/main/java/org/jboss/arquillian/test/spi/execution/ExecutionDecision.java]: sure

**Comment by aslakknutsen** [comment]
Why do we have a TestExecutionDecider.precedence here?

All TestExecutionDecider are executed and we stop with the first one that says DontExecute. So precedence really only seems like an attempt to short curcuit the loop early saying something like 'I'm very likely to stop this from executing so try me first' ?

I've got this code squashed etc in a local branch, so no need to commit more here now..

**Comment by aslakknutsen** [comment]
@smiklosovic It comes out the same in the report if the Decision is done in-container. The difference is that it comes back to the Client side as AssumptionViolation with Status Skipped since it's been converted in-container from a SkippedException already

**Comment by smiklosovic** [comment]
@aslakknutsen commenting on that precedence thing - the scenario you are descibing is behing that. I am not sure if we ever really need that but imagine you have mupltiple deciders and some decider is more time / resource consuming that the other one. So being sure that you have to evenutally do that heavy check, lets check the more easier check and when that does not pass, you save yourself a lot of time.

**Comment by aslakknutsen** [comment]
Yeah, I can see that if the deciders come from the same module. If from different modules we're just creating a odd dependency between them.

I'll leave it in for now

**Comment by aslakknutsen** [comment]
Thank you!  :)

Pushed upstream https://github.com/arquillian/arquillian-core/commit/f4fef41f8fe812af73762f389d4f3ed31f5e47ac

**Comment by smiklosovic** [comment]
thanks!

**Pull request closed without merging**